### PR TITLE
calidns: add --increment and --want-recursion flags

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
This will allow using e.g. a stable amount of QPS (`--increment 1`) and test recursors.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled and tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/calidns.cc [hunk 1]
```diff
@@ -181,34 +186,54 @@ try
...
   struct sched_param param;
   param.sched_priority=99;
+  int arg_i = 1;
+  int increment = 1.1;
+  bool wantRecursion = false;
 
-  if (argc == 1 || (argc > 1 && argc <5)) {
-    for(int i = 1; i<argc; i++) {
-      string opt(argv[i]);
+  if (argc < 5) {
```
Comment (Habbie) [inline pdns/calidns.cc]: This seems redundant with new line 224 below

File: pdns/calidns.cc [hunk 2]
```diff
@@ -181,34 +186,54 @@ try
 {
   struct sched_param param;
   param.sched_priority=99;
+  int arg_i = 1;
```
Comment (Habbie) [inline pdns/calidns.cc]: `argv++; argc--;` after every consumed option simplifies the code and makes `arg_i` obsolete

File: pdns/calidns.cc [hunk 3]
```diff
@@ -181,34 +186,54 @@ try
...
+  int increment = 1.1;
```
Comment (Habbie) [inline pdns/calidns.cc]: int 1.1?

File: pdns/calidns.cc [hunk 4]
```diff
@@ -181,34 +186,54 @@ try
...
 
-    usage();
-    if (argc == 1)
+    if(opt == "--version") {
+      cerr<<"calidns "<<VERSION<<endl;
       exit(EXIT_SUCCESS);
+    }
+
+    if(opt == "--increment") {
+      increment = std::stoi(argv[arg_i+1]);
```
Comment (Habbie) [inline pdns/calidns.cc]: stoi is for ints, increment wants to be float

**Comment by pieterlexis** [comment]
@Habbie can you re-review?

File: pdns/calidns.cc [hunk 5]
```diff
@@ -179,36 +184,69 @@ void usage() {
...
+    ("help,h", "Show this helpful message")
+    ("version", "Show the version number")
+    ("increment", po::value<float>()->default_value(1.1),  "Set the factor to increase the QPS load per run")
+    ("want-recursion", "Set the Recursion Desired flag on queries");
+  po::options_description alloptions;
+  po::options_description hidden("hidden options");
+  hidden.add_options()
+    ("query-file", po::value<string>(), "File with queries")
+    ("destination", po::value<string>(), "Destination address")
+    ("initital-qps", po::value<uint32_t>(), "Initial number of queries per second")
```
Comment (Habbie) [inline pdns/calidns.cc]: initital

File: docs/manpages/calidns.1.md [hunk 6]
```diff
@@ -28,4 +28,9 @@ google.com A
 This is similar to Alexa top 1 million list.
 
 # OPTIONS
-None
+--increment *NUM*
+:    On every subsequent run, multiply the number of queries per second by *NUM*.
+     By default, this is 1.1.
+
+--want-recursion
+:    Set this flag to send queries with the Recursion Desired flag set.
```
Comment (Habbie) [inline docs/manpages/calidns.1.md]: initial-qps missing here?

File: pdns/calidns.cc [hunk 7]
```diff
@@ -241,9 +286,9 @@ try
   int qps;
 
   ofstream plot("plot");
-  for(qps=qpsstart;;qps *= 1.1) {
+  for(qps=qpsstart;;qps *= increment) {
     double seconds=1;
-    cout<<"Aiming at "<<qps<< "qps for "<<seconds<<" seconds at cache hitrate "<<100.0*hitrate<<"%";
+    cout<<"Aiming at "<<qps<< "qps (RD="<<(wantRecursion ? "1" : "0")<<") for "<<seconds<<" seconds at cache hitrate "<<100.0*hitrate<<"%";
```
Comment (Habbie) [inline pdns/calidns.cc]: I think you can just write `<<wantRecursion<<` here?

File: docs/manpages/calidns.1.md (hunk 6 above)
Comment (pieterlexis) [inline docs/manpages/calidns.1.md]: initial-qps is not a flag but a positional option

Comment (Habbie) [inline docs/manpages/calidns.1.md]: oh yes

**Pull request merged**